### PR TITLE
Add ContentRedirectContentGuard

### LINKED ISSUE #2151: Copy redirecting content guard ftom pulp_container to pulpcore
https://github.com/pulp/pulp_ansible/pull/798#discussion_r789478158

### PR DESCRIPTION
**Pull request by mdellweg**
This content guard will use an autogenerated secret to generate
preauthenticated redirect urls to content in guarded distributions.

fixes #2151

### CODE REVIEW

File: pulpcore/app/serializers/publication.py [hunk 1]
```diff
@@ -112,6 +112,33 @@ class RBACContentGuardPermissionSerializer(serializers.Serializer):
     groupnames = serializers.ListField(default=[])
 
 
+class ContentRedirectContentGuardSerializer(ContentGuardSerializer):
+    """
+    A serializer for ContentRedirectContentGuard.
+    """
+
+    @classmethod
+    def get_or_create(cls, natural_key, default_values=None):
```
Comment (ipanova) [inline pulpcore/app/serializers/publication.py:121]: does it make sense to make this available as a class _GetOrCreateMixin, pulp_container uses this in few places.

Comment (mdellweg) [reply pulpcore/app/serializers/publication.py:121]: I was thinking about this, and had no clear answer.

**Comment by mdellweg** [comment]
fyi @newswangerd

File: pulpcore/app/serializers/publication.py (hunk 1 above)
Comment (ipanova) [reply pulpcore/app/serializers/publication.py:123]: i would do that and expose to the plugin so pulp_container can use it and not have code duplicated in core and container https://github.com/pulp/pulp_container/blob/main/pulp_container/app/serializers.py#L141

**Comment by bmbouter** [comment]
Can a functional test be added for this possibly? My hope is that all new features come with functional tests.

**Comment by mdellweg** [comment]
> Can a functional test be added for this possibly? My hope is that all new features come with functional tests.

I do not know, how. There is no code in pulpcore / pulp_file using it.

**Comment by bmbouter** [comment]
Mmmm I agree then, there isn't an opportunity to add a functional test here. @newswangerd maybe your galaxy refactor PR could add this functional test.

@mdellweg what about a unit test for the ContentGuard methods themselves?

**Comment by bmbouter** [comment]
@newswangerd can you confirm this PR is working for you after refactoring your pulp_ansible API PR to use it?

**Comment by newswangerd** [comment]
@mdellweg @bmbouter I can test it.

File: pulpcore/tests/unit/test_content_guard.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        self.assertEqual(url_match2.group("url"), original_url)
+        salt2 = url_match2.group("salt")
+        digest2 = url_match2.group("digest")
+
+        request = Mock()
+
+        # Try unsigned url
+        request.url = original_url
+        request.query = {}
+        with self.assertRaises(PermissionError):
```
Comment (ipanova) [inline pulpcore/tests/unit/test_content_guard.py:40]: nice

**Review by ipanova — APPROVED** [review]

**Comment by bmbouter** [comment]
This all looks good to me. I was hoping for a confirmation from @newswangerd that he can use this in his PR prior to merging, but we need to see if he can focus on that or if we should just merge.

**Comment by newswangerd** [comment]
@bmbouter I've tested this manually and it works. Will add functional tests in pulp ansible along with my PR. I think you can go ahead and merge it now if you want.

File: CHANGES/plugin_api/2151.feature [hunk 3]
```diff
@@ -0,0 +1,4 @@
+Added a redirecting content guard that can be employed by plugins to generate preauthenticated URLs
+that forward from a REST call to the content app.
+Added the ``GetOrCreateSerializerMixin`` to ``get_or_create`` objects still validating them through
```
Comment (bmbouter) [inline CHANGES/plugin_api/2151.feature:3]: nitpick: Can this paragraph of text be consolidated to not be line wrapped at the sentence end?

**Review by bmbouter — APPROVED** [review]
This looks great. I've left one nitpick request. Feel free to merge as you want.

**Pull request merged**